go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -0,0 +1,112 @@
+package doltdb
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/ref"
+	"github.com/dolthub/dolt/go/store/datas"
+)
+
+func Replicate(ctx context.Context, srcDB datas.Database, destDB *DoltDB, tempTableDir string, ds datas.Dataset) error {
+	stRef, ok, err := ds.MaybeHeadRef()
+	t, _ := stRef.TargetType()
+	fmt.Print(t)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		panic("max fix")
+	}
+
+	rf, err := ref.Parse(ds.ID())
+	if err != nil {
+		return err
+	}
+	newCtx, cancelFunc := context.WithCancel(ctx)
+	wg, progChan, pullerEventCh := runProgFuncs(newCtx)
+	puller, err := datas.NewPuller(ctx, tempTableDir, defaultChunksPerTF, srcDB, destDB.db, stRef.TargetHash(), pullerEventCh)
+
+	if err == datas.ErrDBUpToDate {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	err = puller.Pull(ctx)
+	if err != nil {
+		return err
+	}
+
+	stopProgFuncs(cancelFunc, wg, progChan, pullerEventCh)
+	if err != nil {
+		return err
+	}
+
+	err = destDB.SetHead(ctx, rf, stRef)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func pullerProgFunc(ctx context.Context, pullerEventCh <-chan datas.PullerEvent) {
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
the golang book says a switch with multiple available channels will choose equally between them. i've seen the same pattern in the golang source code -- check for done on its own to avoid a case where we have 50/50 chance between being terminating or continuing to consume from the channel. i'm not sure it makes that big of a difference, so i can remove if that explanation doesn't make sense to you.

Reply:
Especially since we don’t do anything at all with the result from the non-done channel, I think foregoing it is fine. With contexts in particular, just checking `ctx.Err() == nil` before going into the select might be clearer/syntactically cleaner.